crates/sui-graphql-e2e-tests/tests/consistency/dynamic_fields.move

```diff
@@ -0,0 +1,549 @@
+        count: u64,
+    }
+
+    public entry fun parent(recipient: address, ctx: &mut TxContext) {
+        transfer::public_transfer(
+            Parent { id: object::new(ctx) },
+            recipient
+        )
+    }
+
+    public entry fun child(recipient: address, ctx: &mut TxContext) {
+        transfer::public_transfer(
+            Child { id: object::new(ctx), count: 0 },
+            recipient
+        )
+    }
+
+    public fun add_child(parent: &mut Parent, child: Child, name: u64) {
+        ofield::add(&mut parent.id, name, child);
+    }
+
+    public fun mutate_child(child: &mut Child) {
+        child.count = child.count + 1;
+    }
+
+    public fun mutate_child_via_parent(parent: &mut Parent, name: u64) {
+        mutate_child(ofield::borrow_mut(&mut parent.id, name))
+    }
+
+    public fun reclaim_child(parent: &mut Parent, name: u64): Child {
+        ofield::remove(&mut parent.id, name)
+    }
+
+    public fun delete_child(parent: &mut Parent, name: u64) {
+        let Child { id, count: _ } = reclaim_child(parent, name);
+        object::delete(id);
+    }
+
+    public entry fun add_df(obj: &mut Parent) {
+        let id = &mut obj.id;
+        field::add<String, String>(id, utf8(b"df1"), utf8(b"df1"));
+        field::add<String, String>(id, utf8(b"df2"), utf8(b"df2"));
+        field::add<String, String>(id, utf8(b"df3"), utf8(b"df3"));
+    }
+
+    public entry fun remove_df(obj: &mut Parent) {
+        let id = &mut obj.id;
+        field::remove<String, String>(id, utf8(b"df1"));
+        field::remove<String, String>(id, utf8(b"df2"));
+        field::remove<String, String>(id, utf8(b"df3"));
+    }
+
+    public entry fun add_more_df(obj: &mut Parent) {
+        let id = &mut obj.id;
+        field::add<String, String>(id, utf8(b"df4"), utf8(b"df4"));
+        field::add<String, String>(id, utf8(b"df5"), utf8(b"df5"));
+        field::add<String, String>(id, utf8(b"df6"), utf8(b"df6"));
+    }
+}
+
+//# programmable --sender A --inputs @A 42
+//> 0: Test::M1::parent(Input(0));
+//> 1: Test::M1::child(Input(0));
+
+//# view-object 2,1
+
+//# view-object 2,0
+
+//# programmable --sender A --inputs object(2,1) object(2,0) 420
+//> Test::M1::add_child(Input(0), Input(1), Input(2));
+//> Test::M1::mutate_child_via_parent(Input(0), Input(2));
+
+//# view-object 2,1
+
+//# view-object 2,0
+
+//# create-checkpoint
+
+//# run-graphql
+fragment DynamicFieldSelect on DynamicField {
+  name {
+    bcs
+    type {
+      repr
+    }
+  }
+  value {
+    ... on MoveObject {
+      contents {
+        json
+      }
+    }
+    ... on MoveValue {
+      json
+    }
+  }
+}
+
+fragment DynamicFieldsSelect on DynamicFieldConnection {
+  edges {
+    cursor
+    node {
+      ...DynamicFieldSelect
+    }
+  }
+}
+
+fragment ParentSelect on Parent {
+  parent {
+    dynamicFields {
+      ...DynamicFieldsSelect
+    }
+  }
+}
+
+{
+  parent_version_2_no_dof_but_will_show_dof: object(address: "@{obj_2_1}", version: 2) {
```

Comment:
This behaviour is not correct.  You correct it by bounding the version of the dynamic field by the version of the parent object. E.g. to get the dynamic fields of a parent at version `v`, we look for the entries in `object_history` + `object_snapshot` that look like dynamic fields, and whose versions are `<= v`.

Reply:
Ah, thanks for catching this